src/wasm/wasm-debug.cpp

```diff
@@ -417,6 +435,66 @@ static void updateDebugLines(const Module& wasm,
   }
 }
 
+// Iterate in parallel over a DwarfContext representation element and a
+// YAML element, which parallel each other.
+template<typename T, typename U, typename W>
+static void iterContextAndYAML(const T& contextList, U& yamlList, W func) {
+  auto yamlValue = yamlList.begin();
+  for (const auto& contextValue : contextList) {
+    assert(yamlValue != yamlList.end());
+    func(contextValue, *yamlValue);
+    yamlValue++;
+  }
+  assert(yamlValue == yamlList.end());
+}
+
+static void updateCompileUnits(const BinaryenDWARFInfo& info,
+                               llvm::DWARFYAML::Data& yaml,
+                               const LocationUpdater& locationUpdater) {
+  // The context has the high-level information we need, and the YAML is where
+  // we write changes. First, iterate over the compile units.
+  iterContextAndYAML(
+    info.context->compile_units(),
+    yaml.CompileUnits,
+    [&](const std::unique_ptr<llvm::DWARFUnit>& CU,
+        llvm::DWARFYAML::Unit& yamlUnit) {
+      // Process the DIEs in each compile unit.
+      iterContextAndYAML(
+        CU->dies(),
+        yamlUnit.Entries,
+        [&](const llvm::DWARFDebugInfoEntry& DIE,
+            llvm::DWARFYAML::Entry& yamlEntry) {
+          // Process the entries in each relevant DIE, looking for attributes to
+          // change.
+          auto abbrevDecl = DIE.getAbbreviationDeclarationPtr();
+          if (abbrevDecl) {
+            iterContextAndYAML(
+              abbrevDecl->attributes(),
+              yamlEntry.Values,
+              [&](const llvm::DWARFAbbreviationDeclaration::AttributeSpec&
+                    attrSpec,
+                  llvm::DWARFYAML::FormValue& yamlValue) {
+                if (attrSpec.Attr == llvm::dwarf::DW_AT_low_pc) {
+                  // If the old address did not refer to an instruction, then
+                  // this is not something we understand and can update.
+                  if (locationUpdater.hasOldAddr(yamlValue.Value)) {
+                    // A compile unit's address is not the address of an
+                    // instruction or a function.
```

Comment:
A compile unit's address is the address of a function if there's only one function in the CU (it has a low_pc/high_pc instead of ranges). The bit currently is that the function's address is also not an instruction.

Reply:
Thanks, yes, that was poorly worded. Improved now I hope...